### DISCUSSION TITLE
Example of executing via embedded library?

### DISCUSSION THREAD

**Discussion by tigerinus** [deepnoodle-ai/risor#10] [Q&A] [answered]
This project looks interesting. 

Is there an example of running a `.tm` from within a custom Golang code?

Thanks.

**Comment by myzie** [comment, marked as answer]
Hi @tigerinus, thanks for dropping by. Here are a couple examples that are relevant, although please let me know if this doesn't fully answer your question.

1. https://github.com/cloudcmds/tamarin/blob/main/cmd/simple-example/main.go
2. https://github.com/cloudcmds/tamarin/blob/main/tamarin.go#L70-L85

The first link there is meant to look very similar to what you would need to do in your own Go program.

**Reply by tigerinus** [reply]
This is great.

My extra need from your scripting language is some kind of interoperability with existing Go code, contracted by interface for example. I am not sure how difficult will this be, but it's going to be a big deal if it happens.

**Reply by myzie** [reply]
Do you specifically need to make calls from Tamarin back into Go functions that you write? This is possible today by adding your own built-in functions using this mechanism:
https://github.com/cloudcmds/tamarin/blob/main/exec/exec.go#L70

However I don't have a super concise example of how to do that, so I should probably put that together.

**Reply by tigerinus** [reply]
Yeah, that built-in function mechanism is helpful.

Can we go extra mile by making it (semi-) runtime based call, versus hard coded in built-in function?

I don't mind a `go:generate` approach to pre-scan the `.go` file with `interface` structs in it, and pre-populate build-in functions based on the scan.

**Reply by myzie** [reply]
That is a fun and interesting idea to use go:generate to help with this. I think it would help make this concrete if you could share an example of an interface you'd want to expose to the Tamarin code. 

I'm guessing you want the interface to deal with native Go types, rather than Tamarin object types? But I want to check on that.

As in, you want your interface signatures to be something other than this, the built-in function signature:

```go
type BuiltinFunction func(ctx context.Context, args ...Object) Object
```

(from https://github.com/cloudcmds/tamarin/blob/main/object/builtin.go#L9)

**Reply by tigerinus** [reply]
Here is a concrete example from my project:

https://github.com/IceWhaleTech/CasaOS/blob/main/service/notify.go

It'd be beautiful to allow my developers to script certain functionalities against the `NotifyServer` interface, load and run them on the fly. That greatly reduces the development cost for 3rd party contributors. It also makes the idea of script based extensions possible.

**Reply by myzie** [reply]
Perfect, thanks for this example. Transforming the input and output types seems like the most difficult aspect of this, and it seems like we could get it to work in a lot of situations. I'll look into this.

**Reply by myzie** [reply]
@tigerinus - check out the changes in [v0.0.9](https://github.com/cloudcmds/tamarin/releases/tag/v0.0.9). [Here is an example](https://github.com/cloudcmds/tamarin/blob/main/cmd/example-proxy/main.go) of proxying to a Go service. This may work for your `NotifyServer` use case, if you register the handful of types you have with a Tamarin ProxyManager.

**Comment by andrewkroh** [comment]
I was looking over the examples. I think it would be useful to be able to parse once and then reuse the returned `ast.Program` with repeated evaluations. It looks like that's not quite possible for external users because the parser is in `internal/`? I am thinking of a similar use case as @tigerinus where the script is an extension point in an app, and you load it once and then execute it repeatedly upon some event to customize behavior.

**Reply by myzie** [reply]
Yeah. It looks like if I pull the `internal/parser` package out of `internal` then that would be supported. I'll make this update soon.

**Reply by myzie** [reply]
Hi @andrewkroh - I made these updates. Here is an example of reusing an already parsed AST:
https://github.com/cloudcmds/tamarin/blob/main/cmd/example-ast-reuse/main.go

**Reply by andrewkroh** [reply]
Thanks! I'll try that out.
